jgrapht-core/src/main/java/org/jgrapht/alg/clustering/KSpanningTreeClustering.java

```diff
@@ -0,0 +1,119 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR LGPL-2.1-or-later
+ */
+package org.jgrapht.alg.clustering;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.jgrapht.Graph;
+import org.jgrapht.GraphTests;
+import org.jgrapht.alg.interfaces.ClusteringAlgorithm;
+import org.jgrapht.alg.interfaces.SpanningTreeAlgorithm.SpanningTree;
+import org.jgrapht.alg.spanning.PrimMinimumSpanningTree;
+import org.jgrapht.alg.util.UnionFind;
+
+/**
+ * The k spanning tree clustering algorithm.
+ * 
+ * <p>
+ * The algorithm finds a minimum spanning tree $T$ using Prim's algorithm, then executes Kruskal's
+ * algorithm only on the edges of $T$ until $k$ trees are formed. The resulting trees are the final
+ * clusters. The total running time is $O(m + n \log n)$.
+ * 
+ * <p>
+ * The algorithm is strongly related to single linkage cluster analysis, also known as single-link
+ * clustering. For more information see: J. C. Gower and G. J. S. Ross. Minimum Spanning Trees and
+ * Single Linkage Cluster Analysis. Journal of the Royal Statistical Society. Series C (Applied
+ * Statistics), 18(1):54--64, 1969.
+ * 
+ * @author Dimitrios Michail
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ */
+public class KSpanningTreeClustering<V, E>
+    implements
+    ClusteringAlgorithm<V>
+{
+    private Graph<V, E> graph;
+    private int k;
+
+    /**
+     * Create a new clustering algorithm.
+     * 
+     * @param graph the graph (needs to be undirected)
+     * @param k the desired number of clusters
+     */
+    public KSpanningTreeClustering(Graph<V, E> graph, int k)
+    {
+        this.graph = GraphTests.requireUndirected(graph);
+        if (k < 1 || k > graph.vertexSet().size()) {
+            throw new IllegalArgumentException("Illegal number of clusters");
+        }
+        this.k = k;
+    }
+
+    @Override
+    public Clustering<V> getClustering()
+    {
+        /*
+         * Compute an MST
+         */
+        SpanningTree<E> mst = new PrimMinimumSpanningTree<>(graph).getSpanningTree();
+
+        /*
+         * Run Kruskal only on MST edges until we get k clusters
+         */
+        UnionFind<V> forest = new UnionFind<>(graph.vertexSet());
+        ArrayList<E> allEdges = new ArrayList<>(mst.getEdges());
+        allEdges.sort(Comparator.comparingDouble(graph::getEdgeWeight));
+
+        for (E edge : allEdges) {
+            if (forest.numberOfSets() == k) {
+                break;
+            }
+            V source = graph.getEdgeSource(edge);
+            V target = graph.getEdgeTarget(edge);
+            if (forest.find(source).equals(forest.find(target))) {
+                continue;
+            }
+
+            forest.union(source, target);
+        }
+
+        /*
+         * Transform and return result
+         */
+        Map<V, Set<V>> clusterMap = new HashMap<>();
```

Comment:
Shouldn't this be a LinkedHashMap?

Reply:
Now that you mention it, the other ones should (HashSet -> LinkedHashSet). This one is used only for the grouping, so its order does not matter.

Reply:
Or maybe you are right, so that we always return the same ordering of the sets.